### ISSUE TITLE
6.1.1 missing wsgiserver install regression?

### ISSUE THREAD

**Issue by pcmillerAcme** [cherrypy/cherrypy#1456]
Hi friends - If I upgrade to cherrypy-6.1.1 from cherrpy-6.0.1 my service breaks. CP launches with new errors about 'no wsgiserver' There are several pages online showing I'm not alone in this, several talking about having to manually installing wsgiserver.py from some strange place to make cherrpy-6.1.1 work or putting in files to make some process force wsgi server files to arrive. I've tried multiple methods to put in a wsgiserver to work with cherrypy-6.1.1, but either it errors or fails to serve (never responds)

Any pointers? I'm unclear on what is posted here and how much that correlates with what one gets by 

`pip install cherrypy --upgrade`

Happily, I can regress to cherrypy-6.0.1 and everyhing is ok:

`pip.exe install cherrypy==6.0.1`

I'm on windows running python-2.7.11

Happy to provide any more details if that would help.

**Comment by jaraco** [comment]
I tried to replicate the failure on a clean environment, but cannot.

```
$ python -m venv ~/.envs/issue-1456
$ ~/.envs/issue-1456/bin/python -m pip install cherrypy
Collecting cherrypy
  Downloading CherryPy-6.1.1-py3-none-any.whl (457kB)
    100% |████████████████████████████████| 460kB 2.2MB/s 
Collecting six (from cherrypy)
  Using cached six-1.10.0-py2.py3-none-any.whl
Installing collected packages: six, cherrypy
Successfully installed cherrypy-6.1.1 six-1.10.0
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
$ ~/.envs/issue-1456/bin/python ~/p/cherrypy/cherrypy/tutorial/tut01_helloworld.py
[17/Jul/2016:21:30:20] ENGINE Listening for SIGTERM.
[17/Jul/2016:21:30:20] ENGINE Listening for SIGUSR1.
[17/Jul/2016:21:30:20] ENGINE Listening for SIGHUP.
[17/Jul/2016:21:30:20] ENGINE Bus STARTING
[17/Jul/2016:21:30:20] ENGINE Started monitor thread '_TimeoutMonitor'.
[17/Jul/2016:21:30:20] ENGINE Started monitor thread 'Autoreloader'.
[17/Jul/2016:21:30:21] ENGINE Serving on http://127.0.0.1:8080
[17/Jul/2016:21:30:21] ENGINE Bus STARTED
127.0.0.1 - - [17/Jul/2016:21:30:29] "GET / HTTP/1.1" 200 12 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
127.0.0.1 - - [17/Jul/2016:21:30:29] "GET /favicon.ico HTTP/1.1" 404 1478 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
127.0.0.1 - - [17/Jul/2016:21:30:29] "GET /favicon.ico HTTP/1.1" 404 1478 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
127.0.0.1 - - [17/Jul/2016:21:30:29] "GET /favicon.ico HTTP/1.1" 404 1478 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
^C[17/Jul/2016:21:30:47] ENGINE Keyboard Interrupt: shutting down bus
[17/Jul/2016:21:30:47] ENGINE Bus STOPPING
[17/Jul/2016:21:30:47] ENGINE HTTP Server cherrypy._cpwsgi_server.CPWSGIServer(('127.0.0.1', 8080)) shut down
[17/Jul/2016:21:30:47] ENGINE Stopped thread '_TimeoutMonitor'.
[17/Jul/2016:21:30:47] ENGINE Stopped thread 'Autoreloader'.
[17/Jul/2016:21:30:47] ENGINE Bus STOPPED
[17/Jul/2016:21:30:47] ENGINE Bus EXITING
[17/Jul/2016:21:30:47] ENGINE Bus EXITED
[17/Jul/2016:21:30:47] ENGINE Waiting for child threads to terminate...
$ python -m virtualenv --python python2.7 ~/.virtualenvs/issue-1456
Running virtualenv with interpreter /usr/bin/python2.7
New python executable in /Users/jaraco/.virtualenvs/issue-1456/bin/python
Installing setuptools, pip, wheel...done.
$ ~/.virtualenvs/issue-1456/bin/python -m pip install cherrypy
Collecting cherrypy
  Downloading CherryPy-6.1.1.tar.gz (415kB)
    100% |████████████████████████████████| 419kB 2.2MB/s 
Collecting six (from cherrypy)
  Using cached six-1.10.0-py2.py3-none-any.whl
Building wheels for collected packages: cherrypy
  Running setup.py bdist_wheel for cherrypy ... done
  Stored in directory: /Users/jaraco/Library/Caches/pip/wheels/ae/05/58/f7bd0b7bfd16b8ddee9e5431ee6d09d01aa3805d3108735181
Successfully built cherrypy
Installing collected packages: six, cherrypy
Successfully installed cherrypy-6.1.1 six-1.10.0
$ ~/.virtualenvs/issue-1456/bin/python ~/p/cherrypy/cherrypy/tutorial/tut01_helloworld.py
[17/Jul/2016:21:31:47] ENGINE Listening for SIGHUP.
[17/Jul/2016:21:31:47] ENGINE Listening for SIGTERM.
[17/Jul/2016:21:31:47] ENGINE Listening for SIGUSR1.
[17/Jul/2016:21:31:47] ENGINE Bus STARTING
[17/Jul/2016:21:31:47] ENGINE Started monitor thread 'Autoreloader'.
[17/Jul/2016:21:31:47] ENGINE Started monitor thread '_TimeoutMonitor'.
[17/Jul/2016:21:31:47] ENGINE Serving on http://127.0.0.1:8080
[17/Jul/2016:21:31:47] ENGINE Bus STARTED
127.0.0.1 - - [17/Jul/2016:21:31:50] "GET / HTTP/1.1" 200 12 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
```

I notice you say that the issue occurs if you upgrade from 6.0.1. I'll try that next.

**Comment by jaraco** [comment]
Even after a downgrade and upgrade, the service starts up and responds exactly as expected.

```
$ ~/.virtualenvs/issue-1456/bin/python -m pip install cherrypy==6.0.1                    
Collecting cherrypy==6.0.1
  Downloading CherryPy-6.0.1.tar.gz (435kB)
    100% |████████████████████████████████| 440kB 2.6MB/s 
Requirement already satisfied (use --upgrade to upgrade): six in ./.virtualenvs/issue-1456/lib/python2.7/site-packages (from cherrypy==6.0.1)
Building wheels for collected packages: cherrypy
  Running setup.py bdist_wheel for cherrypy ... done
  Stored in directory: /Users/jaraco/Library/Caches/pip/wheels/83/83/b2/22963132091bbb65bbfa99e41cb2440e8688cbfed44ce3369f
Successfully built cherrypy
Installing collected packages: cherrypy
  Found existing installation: CherryPy 6.1.1
    Uninstalling CherryPy-6.1.1:
      Successfully uninstalled CherryPy-6.1.1
Successfully installed cherrypy-6.0.1
$ ~/.virtualenvs/issue-1456/bin/python -m pip install cherrypy==6.1.1
Collecting cherrypy==6.1.1
Requirement already satisfied (use --upgrade to upgrade): six in ./.virtualenvs/issue-1456/lib/python2.7/site-packages (from cherrypy==6.1.1)
Installing collected packages: cherrypy
  Found existing installation: CherryPy 6.0.1
    Uninstalling CherryPy-6.0.1:
      Successfully uninstalled CherryPy-6.0.1
Successfully installed cherrypy-6.1.1
$ ~/.virtualenvs/issue-1456/bin/python ~/p/cherrypy/cherrypy/tutorial/tut01_helloworld.py
[17/Jul/2016:21:34:47] ENGINE Listening for SIGUSR1.
[17/Jul/2016:21:34:47] ENGINE Listening for SIGHUP.
[17/Jul/2016:21:34:47] ENGINE Listening for SIGTERM.
[17/Jul/2016:21:34:47] ENGINE Bus STARTING
[17/Jul/2016:21:34:47] ENGINE Started monitor thread '_TimeoutMonitor'.
[17/Jul/2016:21:34:47] ENGINE Started monitor thread 'Autoreloader'.
[17/Jul/2016:21:34:47] ENGINE Serving on http://127.0.0.1:8080
[17/Jul/2016:21:34:47] ENGINE Bus STARTED
127.0.0.1 - - [17/Jul/2016:21:34:50] "GET /favicon.ico HTTP/1.1" 404 1488 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
127.0.0.1 - - [17/Jul/2016:21:34:52] "GET / HTTP/1.1" 200 12 "" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.11; rv:47.0) Gecko/20100101 Firefox/47.0"
```

**Comment by jaraco** [comment]
> Happy to provide any more details if that would help.

Yes, please. Can you send the error message you get? When this happens, can you inspect your Python environment to see what cherrypy artifacts exist, especially around wsgiserver?

The substantial change between CP 6.0 and 6.1 was the combining of `cherrypy.wsgiserver.wsgiserver3` and `cherrypy.wsgiserver.wsgiserver2` into simply `cherrypy.wsgiserver` (`__init__.py`), which was the place where the contents of wsgiserver were copied anyway, so the change should be compatible. Please help me understand your case and how you've found a case where this change is apparently not compatible.

It's quite possible that the issue is with the installer and not with the CP package itself.

**Comment by pcmillerAcme** [comment]
Jason - 

Thanks very much! With you clarification on the changes for wsgiserver, I was able to change my code from
`from cherrypy.wsgiserver.wsgiserver2 import ssl_adapters`
to
`from cherrypy.wsgiserver import ssl_adapters`
to be cherrypy-6.1.1 compatible and all was well. Much appreciated!

Side note: I have this long term sense of frustration of not being able to find docs on just the delta between versions. I found the CHANGES.txt file this time, but... feels like a 'possible required code changes' file for versions would be handy. But I dream.... more security for software engineers I guess. Now on to CVE-2016-2107. :-)

Thanks again for all your attention to this and have a good evening!

-Peter
